document.applets.length is O(n)

rniwa opened the pull request:
#### db245ccc887cfb36524423078e1f7b4fbeef55b7
<pre>
document.applets.length is O(n)
<a href="https://bugs.webkit.org/show_bug.cgi?id=255389">https://bugs.webkit.org/show_bug.cgi?id=255389</a>

Reviewed by Chris Dumez.

Introduce and deploy EmptyHTMLCollection, which is always empty, for document.applets.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::applets):
* Source/WebCore/html/EmptyHTMLCollection.h: Added.
* Source/WebCore/html/GenericCachedHTMLCollection.cpp:
(WebCore::GenericCachedHTMLCollection&lt;traversalType&gt;::elementMatches const):

Canonical link: <a href="https://commits.webkit.org/262911@main">https://commits.webkit.org/262911@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd86b3be29ea8a0fbd3c774e5945804919f69f17

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2994 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3058 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3161 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4403 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3389 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2961 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3137 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3099 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2618 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3024 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3382 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2685 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4197 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/879 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2665 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2512 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2653 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2716 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3933 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3062 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2455 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2691 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2667 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/738 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2679 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2882 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez reviewed (commented):
This is about as much code as my EmptyHTMLCollection proposal. Also, the document.applets collection will still have an unnecessary CollectionIndexCache. It will also still register with the document for invalidation and get calls for invalidation when the DOM tree changes AFAICT.

Could you clarify why this is better?

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -6294,7 +6295,7 @@ Ref<HTMLCollection> Document::images()
 Ref<HTMLCollection> Document::applets()
 {
-    return ensureCachedCollection<DocEmpty>();
+    return ensureRareData().ensureNodeLists().addCachedCollection<EmptyHTMLCollection>(*this, DocEmpty);
```
I think it would be nice to move `DocEmpty` to the ASSERT_NOT_REACHED() section (with `DocAll`) inside
`GenericCachedHTMLCollection<traversalType>::elementMatches(Element&` instead of making it look like GenericCachedHTMLCollection is dealing with this type.

rniwa replied:
Oh, right, I forgot to do that.

cdumez reviewed (approved):
LGTM with nit.

cdumez reviewed (approved):
LGTM with nit.

rniwa committed:
document.applets.length is O(n)
https://bugs.webkit.org/show_bug.cgi?id=255389

Reviewed by Chris Dumez.

Introduce and deploy EmptyHTMLCollection, which is always empty, for document.applets.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::applets):
* Source/WebCore/html/EmptyHTMLCollection.h: Added.
* Source/WebCore/html/GenericCachedHTMLCollection.cpp:
(WebCore::GenericCachedHTMLCollection<traversalType>::elementMatches const):

Canonical link: https://commits.webkit.org/262911@main
